// @progress/kendo-angular-editor/dist/es/tools/editor-fontsize.component.js
import * as tslib_1 from "tslib";
import { Component, TemplateRef, ViewChild, ElementRef, forwardRef, Output, Host, EventEmitter } from '@angular/core';
import { ToolBarToolComponent } from '@progress/kendo-angular-toolbar';
import { outerWidth, isPresent } from '../util';
import { EditorComponent } from '../editor.component';
import { getUniqueStyleValues } from '../prosemirror-utils';
import { EditorLocalizationService } from '../localization/editor-localization.service';
/**
 * A component which configures an existing `DropDownListComponent` as an Editor tool
 * ([see example]({% slug toolbartools_editor %}#toc-built-in-tools)).
 * The component associates a `kendo-dropdownlist` with an Editor command that changes the font size of a content block and
 * automatically defines the options of the drop-down list and sets its values.
 *
 * @example
 * ```ts-no-run
 * <kendo-toolbar-dropdownlist kendoEditorFontSize></kendo-toolbar-dropdownlist>
 * ```
 */
var EditorFontSizeComponent = /** @class */ (function (_super) {
    tslib_1.__extends(EditorFontSizeComponent, _super);
    function EditorFontSizeComponent(editor, localization) {
        var _this = _super.call(this) || this;
        _this.editor = editor;
        _this.localization = localization;
        /**
         * Fires when the user updates the value of the drop-down list.
         */
        _this.valueChange = new EventEmitter();
        return _this;
    }
    EditorFontSizeComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.itemDisabled = function (itemArgs) {
            return itemArgs.dataItem.size === null;
        };
        this.data = [
            { text: '8px', size: '8' },
            { text: '10px', size: '10' },
            { text: '12px', size: '12' },
            { text: '14px', size: '14' },
            { text: '18px', size: '18' },
            { text: '24px', size: '24' },
            { text: '36px', size: '36' }
        ];
        setTimeout(function () { return _this.defaultItem = { text: _this.title, size: null }; });
        this.subs = this.editor.stateChange.subscribe(function (_a) {
            var style = _a.style;
            // remove units(px, em, rem...)
            // string#match returns array
            _this.value = (getUniqueStyleValues(style, 'font-size').match(/\d+/g) || [null])[0];
        });
    };
    /**
     * @hidden
     */
    EditorFontSizeComponent.prototype.onValueChange = function (ev) {
        if (isPresent(ev)) {
            this.editor.exec('fontSize', ev);
            this.editor.focus();
            this.valueChange.emit(this.data.find(function (d) { return d.size === ev; }));
        }
    };
    EditorFontSizeComponent.prototype.ngOnDestroy = function () {
        this.subs.unsubscribe();
    };
    Object.defineProperty(EditorFontSizeComponent.prototype, "outerWidth", {
        get: function () {
            if (this.element) {
                return outerWidth(this.element.nativeElement);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EditorFontSizeComponent.prototype, "title", {
        get: function () {
            return this.localization.get('fontSize');
        },
        enumerable: true,
        configurable: true
    });
    EditorFontSizeComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:no-forward-ref
                    providers: [{ provide: ToolBarToolComponent, useExisting: forwardRef(function () { return EditorFontSizeComponent; }) }],
                    selector: 'kendo-toolbar-dropdownlist[kendoEditorFontSize]',
                    template: "\n        <ng-template #toolbarTemplate>\n            <kendo-dropdownlist\n                #element\n                [defaultItem]=\"defaultItem\"\n                [textField]=\"'text'\"\n                [valueField]=\"'size'\"\n                [data]=\"data\"\n                [(value)]=\"value\"\n                [valuePrimitive]=\"true\"\n                [itemDisabled]=\"itemDisabled\"\n                [attr.title]=\"title\"\n                (valueChange)=\"onValueChange($event)\"\n            >\n            </kendo-dropdownlist>\n        </ng-template>\n    "
                },] },
    ];
    /** @nocollapse */
    EditorFontSizeComponent.ctorParameters = function () { return [
        { type: EditorComponent, decorators: [{ type: Host }] },
        { type: EditorLocalizationService }
    ]; };
    EditorFontSizeComponent.propDecorators = {
        valueChange: [{ type: Output }],
        toolbarTemplate: [{ type: ViewChild, args: ['toolbarTemplate',] }],
        element: [{ type: ViewChild, args: ['element',] }]
    };
    return EditorFontSizeComponent;
}(ToolBarToolComponent));
export { EditorFontSizeComponent };
