// @progress/kendo-angular-editor/dist/es/editor-toolbar-state.js
import { undo, redo } from 'prosemirror-history';
import { alignRightRules, alignCenterRules, alignRemoveRules, alignLeftRules } from './config/align-rules';
import { schema } from './config/schema';
import { isMarkActive, activeNode, canIndent, canOutdent, isInListOfType, isAligned, canBeIndented, getActiveMarks } from './prosemirror-utils';
import { indentRules, outdentRules } from './config/indent-rules';
/**
 * @hidden
 */
export var getToolbarState = function (state) { return ({
    alignCenter: isAligned(state, alignCenterRules),
    alignJustify: isAligned(state, alignRemoveRules),
    alignLeft: isAligned(state, alignLeftRules),
    alignRight: isAligned(state, alignRightRules),
    bold: isMarkActive(state, schema.marks.strong),
    cleanFormatting: state.selection.empty,
    format: activeNode(state),
    indent: canIndent(state, schema.nodes.listItem) || canBeIndented(state, indentRules),
    insertOrderedList: isInListOfType(state, schema.nodes.orderedList),
    insertUnorderedList: isInListOfType(state, schema.nodes.bulletList),
    italic: isMarkActive(state, schema.marks.em),
    link: isMarkActive(state, schema.marks.link),
    noSelection: state.selection.empty,
    outdent: canOutdent(state, schema.nodes.listItem) || canBeIndented(state, outdentRules),
    redo: !redo(state),
    strikethrough: isMarkActive(state, schema.marks.strikethrough),
    style: getActiveMarks(state, schema.marks.style),
    subscript: isMarkActive(state, schema.marks.subscript),
    superscript: isMarkActive(state, schema.marks.superscript),
    underline: isMarkActive(state, schema.marks.underline),
    undo: !undo(state)
}); };
/**
 * @hidden
 */
export var initialToolBarState = {
    //marks
    bold: false,
    italic: false,
    underline: false,
    strikethrough: false,
    unlink: true,
    subscript: false,
    superscript: false,
    style: { marks: [], hasNodesWithoutMarks: false },
    //nodes
    format: { text: 'Format', tag: null },
    insertOrderedList: false,
    insertUnorderedList: false,
    indent: false,
    outdent: false,
    alignLeft: false,
    alignCenter: false,
    alignRight: false,
    alignJustify: false,
    //history
    undo: true,
    redo: true,
    //selection
    noSelection: true,
    cleanFormatting: true
};
