// @progress/kendo-angular-editor/dist/es2015/tools/editor-fontsize.component.js
import { Component, TemplateRef, ViewChild, ElementRef, forwardRef, Output, Host, EventEmitter } from '@angular/core';
import { ToolBarToolComponent } from '@progress/kendo-angular-toolbar';
import { outerWidth, isPresent } from '../util';
import { EditorComponent } from '../editor.component';
import { getUniqueStyleValues } from '../prosemirror-utils';
import { EditorLocalizationService } from '../localization/editor-localization.service';
/**
 * A component which configures an existing `DropDownListComponent` as an Editor tool
 * ([see example]({% slug toolbartools_editor %}#toc-built-in-tools)).
 * The component associates a `kendo-dropdownlist` with an Editor command that changes the font size of a content block and
 * automatically defines the options of the drop-down list and sets its values.
 *
 * @example
 * ```ts-no-run
 * <kendo-toolbar-dropdownlist kendoEditorFontSize></kendo-toolbar-dropdownlist>
 * ```
 */
export class EditorFontSizeComponent extends ToolBarToolComponent {
    constructor(editor, localization) {
        super();
        this.editor = editor;
        this.localization = localization;
        /**
         * Fires when the user updates the value of the drop-down list.
         */
        this.valueChange = new EventEmitter();
    }
    ngOnInit() {
        this.itemDisabled = itemArgs => {
            return itemArgs.dataItem.size === null;
        };
        this.data = [
            { text: '8px', size: '8' },
            { text: '10px', size: '10' },
            { text: '12px', size: '12' },
            { text: '14px', size: '14' },
            { text: '18px', size: '18' },
            { text: '24px', size: '24' },
            { text: '36px', size: '36' }
        ];
        setTimeout(() => this.defaultItem = { text: this.title, size: null });
        this.subs = this.editor.stateChange.subscribe(({ style }) => {
            // remove units(px, em, rem...)
            // string#match returns array
            this.value = (getUniqueStyleValues(style, 'font-size').match(/\d+/g) || [null])[0];
        });
    }
    /**
     * @hidden
     */
    onValueChange(ev) {
        if (isPresent(ev)) {
            this.editor.exec('fontSize', ev);
            this.editor.focus();
            this.valueChange.emit(this.data.find(d => d.size === ev));
        }
    }
    ngOnDestroy() {
        this.subs.unsubscribe();
    }
    get outerWidth() {
        if (this.element) {
            return outerWidth(this.element.nativeElement);
        }
    }
    get title() {
        return this.localization.get('fontSize');
    }
}
EditorFontSizeComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:no-forward-ref
                providers: [{ provide: ToolBarToolComponent, useExisting: forwardRef(() => EditorFontSizeComponent) }],
                selector: 'kendo-toolbar-dropdownlist[kendoEditorFontSize]',
                template: `
        <ng-template #toolbarTemplate>
            <kendo-dropdownlist
                #element
                [defaultItem]="defaultItem"
                [textField]="'text'"
                [valueField]="'size'"
                [data]="data"
                [(value)]="value"
                [valuePrimitive]="true"
                [itemDisabled]="itemDisabled"
                [attr.title]="title"
                (valueChange)="onValueChange($event)"
            >
            </kendo-dropdownlist>
        </ng-template>
    `
            },] },
];
/** @nocollapse */
EditorFontSizeComponent.ctorParameters = () => [
    { type: EditorComponent, decorators: [{ type: Host }] },
    { type: EditorLocalizationService }
];
EditorFontSizeComponent.propDecorators = {
    valueChange: [{ type: Output }],
    toolbarTemplate: [{ type: ViewChild, args: ['toolbarTemplate',] }],
    element: [{ type: ViewChild, args: ['element',] }]
};
