// @progress/kendo-angular-editor/dist/npm/localization/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
/**
 * @hidden
 */
var Messages = /** @class */ (function (_super) {
    tslib_1.__extends(Messages, _super);
    function Messages() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Messages.propDecorators = {
        alignCenter: [{ type: core_1.Input }],
        alignJustify: [{ type: core_1.Input }],
        alignLeft: [{ type: core_1.Input }],
        alignRight: [{ type: core_1.Input }],
        backColor: [{ type: core_1.Input }],
        bold: [{ type: core_1.Input }],
        cleanFormatting: [{ type: core_1.Input }],
        createLink: [{ type: core_1.Input }],
        fontFamily: [{ type: core_1.Input }],
        fontSize: [{ type: core_1.Input }],
        foreColor: [{ type: core_1.Input }],
        format: [{ type: core_1.Input }],
        indent: [{ type: core_1.Input }],
        insertFile: [{ type: core_1.Input }],
        insertImage: [{ type: core_1.Input }],
        insertOrderedList: [{ type: core_1.Input }],
        insertUnorderedList: [{ type: core_1.Input }],
        italic: [{ type: core_1.Input }],
        outdent: [{ type: core_1.Input }],
        redo: [{ type: core_1.Input }],
        strikethrough: [{ type: core_1.Input }],
        subscript: [{ type: core_1.Input }],
        superscript: [{ type: core_1.Input }],
        underline: [{ type: core_1.Input }],
        undo: [{ type: core_1.Input }],
        unlink: [{ type: core_1.Input }],
        viewSource: [{ type: core_1.Input }],
        dialogCancel: [{ type: core_1.Input }],
        dialogInsert: [{ type: core_1.Input }],
        dialogUpdate: [{ type: core_1.Input }],
        fileText: [{ type: core_1.Input }],
        fileTitle: [{ type: core_1.Input }],
        fileWebAddress: [{ type: core_1.Input }],
        imageAltText: [{ type: core_1.Input }],
        imageHeight: [{ type: core_1.Input }],
        imageWebAddress: [{ type: core_1.Input }],
        imageWidth: [{ type: core_1.Input }],
        linkOpenInNewWindow: [{ type: core_1.Input }],
        linkText: [{ type: core_1.Input }],
        linkTitle: [{ type: core_1.Input }],
        linkWebAddress: [{ type: core_1.Input }]
    };
    return Messages;
}(kendo_angular_l10n_1.ComponentMessages));
exports.Messages = Messages;
