// @progress/kendo-angular-scheduler/dist/es/views/common/configuration-view-base.js
import * as tslib_1 from "tslib";
import { TemplateRef, ViewChild, ContentChild, Input } from '@angular/core';
import { SchedulerView } from '../../types';
import { EventTemplateDirective, GroupHeaderTemplateDirective } from '../templates';
import { isPresent } from '../../common/util';
var defaultSlotClass = function (_args) { return null; };
var ɵ0 = defaultSlotClass;
/**
 * @hidden
 */
var ConfigurationViewBase = /** @class */ (function (_super) {
    tslib_1.__extends(ConfigurationViewBase, _super);
    function ConfigurationViewBase(localization, changeDetector, viewContext, viewState) {
        var _this = _super.call(this) || this;
        _this.localization = localization;
        _this.changeDetector = changeDetector;
        _this.viewContext = viewContext;
        _this.viewState = viewState;
        _this.schedulerOptions = {};
        _this.subs = _this.localization.changes.subscribe(function (_a) {
            var rtl = _a.rtl;
            changeDetector.markForCheck();
        });
        _this.subs.add(_this.viewContext.optionsChange.subscribe(_this.optionsChange.bind(_this)));
        return _this;
    }
    Object.defineProperty(ConfigurationViewBase.prototype, "viewSlotClass", {
        /**
         * @hidden
         */
        get: function () {
            return isPresent(this.slotClass) ? this.slotClass : (this.schedulerOptions.slotClass || defaultSlotClass);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConfigurationViewBase.prototype, "viewEventClass", {
        /**
         * @hidden
         */
        get: function () {
            return isPresent(this.eventClass) ? this.eventClass : this.schedulerOptions.eventClass;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConfigurationViewBase.prototype, "viewEventStyles", {
        /**
         * @hidden
         */
        get: function () {
            return isPresent(this.eventStyles) ? this.eventStyles : this.schedulerOptions.eventStyles;
        },
        enumerable: true,
        configurable: true
    });
    ConfigurationViewBase.prototype.ngOnChanges = function (_changes) {
        this.viewState.notifyOptionsChange();
    };
    ConfigurationViewBase.prototype.ngOnDestroy = function () {
        this.subs.unsubscribe();
    };
    ConfigurationViewBase.prototype.optionsChange = function (options) {
        this.schedulerOptions = options;
    };
    ConfigurationViewBase.propDecorators = {
        slotClass: [{ type: Input }],
        eventClass: [{ type: Input }],
        eventStyles: [{ type: Input }],
        template: [{ type: ViewChild, args: ['content',] }],
        eventTemplate: [{ type: ContentChild, args: [EventTemplateDirective,] }],
        groupHeaderTemplate: [{ type: ContentChild, args: [GroupHeaderTemplateDirective,] }]
    };
    return ConfigurationViewBase;
}(SchedulerView));
export { ConfigurationViewBase };
export { ɵ0 };
