// @progress/kendo-angular-editor/dist/es/localization/messages.js
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { ComponentMessages } from '@progress/kendo-angular-l10n';
/**
 * @hidden
 */
var Messages = /** @class */ (function (_super) {
    tslib_1.__extends(Messages, _super);
    function Messages() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Messages.propDecorators = {
        alignCenter: [{ type: Input }],
        alignJustify: [{ type: Input }],
        alignLeft: [{ type: Input }],
        alignRight: [{ type: Input }],
        backColor: [{ type: Input }],
        bold: [{ type: Input }],
        cleanFormatting: [{ type: Input }],
        createLink: [{ type: Input }],
        fontFamily: [{ type: Input }],
        fontSize: [{ type: Input }],
        foreColor: [{ type: Input }],
        format: [{ type: Input }],
        indent: [{ type: Input }],
        insertFile: [{ type: Input }],
        insertImage: [{ type: Input }],
        insertOrderedList: [{ type: Input }],
        insertUnorderedList: [{ type: Input }],
        italic: [{ type: Input }],
        outdent: [{ type: Input }],
        redo: [{ type: Input }],
        strikethrough: [{ type: Input }],
        subscript: [{ type: Input }],
        superscript: [{ type: Input }],
        underline: [{ type: Input }],
        undo: [{ type: Input }],
        unlink: [{ type: Input }],
        viewSource: [{ type: Input }],
        dialogCancel: [{ type: Input }],
        dialogInsert: [{ type: Input }],
        dialogUpdate: [{ type: Input }],
        fileText: [{ type: Input }],
        fileTitle: [{ type: Input }],
        fileWebAddress: [{ type: Input }],
        imageAltText: [{ type: Input }],
        imageHeight: [{ type: Input }],
        imageWebAddress: [{ type: Input }],
        imageWidth: [{ type: Input }],
        linkOpenInNewWindow: [{ type: Input }],
        linkText: [{ type: Input }],
        linkTitle: [{ type: Input }],
        linkWebAddress: [{ type: Input }]
    };
    return Messages;
}(ComponentMessages));
export { Messages };
