// @progress/kendo-angular-scheduler/dist/npm/views/common/configuration-view-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var types_1 = require("../../types");
var templates_1 = require("../templates");
var util_1 = require("../../common/util");
var defaultSlotClass = function (_args) { return null; };
var ɵ0 = defaultSlotClass;
exports.ɵ0 = ɵ0;
/**
 * @hidden
 */
var ConfigurationViewBase = /** @class */ (function (_super) {
    tslib_1.__extends(ConfigurationViewBase, _super);
    function ConfigurationViewBase(localization, changeDetector, viewContext, viewState) {
        var _this = _super.call(this) || this;
        _this.localization = localization;
        _this.changeDetector = changeDetector;
        _this.viewContext = viewContext;
        _this.viewState = viewState;
        _this.schedulerOptions = {};
        _this.subs = _this.localization.changes.subscribe(function (_a) {
            var rtl = _a.rtl;
            changeDetector.markForCheck();
        });
        _this.subs.add(_this.viewContext.optionsChange.subscribe(_this.optionsChange.bind(_this)));
        return _this;
    }
    Object.defineProperty(ConfigurationViewBase.prototype, "viewSlotClass", {
        /**
         * @hidden
         */
        get: function () {
            return util_1.isPresent(this.slotClass) ? this.slotClass : (this.schedulerOptions.slotClass || defaultSlotClass);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConfigurationViewBase.prototype, "viewEventClass", {
        /**
         * @hidden
         */
        get: function () {
            return util_1.isPresent(this.eventClass) ? this.eventClass : this.schedulerOptions.eventClass;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConfigurationViewBase.prototype, "viewEventStyles", {
        /**
         * @hidden
         */
        get: function () {
            return util_1.isPresent(this.eventStyles) ? this.eventStyles : this.schedulerOptions.eventStyles;
        },
        enumerable: true,
        configurable: true
    });
    ConfigurationViewBase.prototype.ngOnChanges = function (_changes) {
        this.viewState.notifyOptionsChange();
    };
    ConfigurationViewBase.prototype.ngOnDestroy = function () {
        this.subs.unsubscribe();
    };
    ConfigurationViewBase.prototype.optionsChange = function (options) {
        this.schedulerOptions = options;
    };
    ConfigurationViewBase.propDecorators = {
        slotClass: [{ type: core_1.Input }],
        eventClass: [{ type: core_1.Input }],
        eventStyles: [{ type: core_1.Input }],
        template: [{ type: core_1.ViewChild, args: ['content',] }],
        eventTemplate: [{ type: core_1.ContentChild, args: [templates_1.EventTemplateDirective,] }],
        groupHeaderTemplate: [{ type: core_1.ContentChild, args: [templates_1.GroupHeaderTemplateDirective,] }]
    };
    return ConfigurationViewBase;
}(types_1.SchedulerView));
exports.ConfigurationViewBase = ConfigurationViewBase;
